### ISSUE TITLE
StaticLambdaFixer: skip known unfixable codeblock

### ISSUE THREAD

**Issue by Slamdunk** [PHP-CS-Fixer/PHP-CS-Fixer#3317]
I tried @SpacePossum new `StaticLambdaFixer` #3187 but can't adopt it because I use Symfony [Command::setCode()](https://github.com/symfony/symfony/blob/v4.0.1/src/Symfony/Component/Console/Command/Command.php#L277) that dinamically binds closures to another object.
Example code:

```php
// $foo = function() {}; // Original code now fixed
$foo = static function() {};

$foo = $foo->bindTo(new \stdClass);
// Warning: Cannot bind an instance to a static closure in /tmp.php on line 3
```

Would it be possible to integrate a comment annotation to ignore specific codeblocks?
Something like:

```php
$foo = function() {}; // Fix this

// @PhpCsFixer::ignore{"static_lambda"}
$foo = function() {}; // Skip this
```

I just made the annotation up; we can borrow Doctrine's one.

**Comment by keradus** [comment]
from rule docs:
> *Risky rule: risky when using "->bindTo" on lambdas without referencing to ``$this``.*

**Comment by keradus** [comment]
currently, this is not possible to say that given line shall be ignored by given rules (or whole tool)

If you want to propose such feature, it would be interesting indeed.

**Comment by SpacePossum** [comment]
Root cause problem is that PHP does not provide a way to test if `bindTo` is possible, which would allow the SF code to test before trying to bind.
I'm not sure how to continue here, if the annotation logic is welcome here by the PHP-CS-Fixer community I'm interested in it as well of course :)

**Comment by Slamdunk** [comment]
After some days of thinking, a general purpose annotation for this functionality seems too much for such a little gain, considering also that the need didn't emerge before.

What about a fixer specific annotation, that would also declare *why* the fixer is skipped?
Something like:

```php
// This will be fixed to static
$foo = function() {
    var_dump('foo');
};

// This will be skipped
$bar = function() {
    // @boundLambda
    var_dump('bar');
};
```

Easy to implement, it would impact only StaticLambdaFixer. WDYT?

**Comment by keradus** [comment]
> What about a fixer specific annotation (...) it would impact only StaticLambdaFixer. WDYT?

That solution won't be accepted by me. Then, when one will need same functionality for other rule, he will recreate that solution. Then, user will want to do sth similar for other rule, and he will discover it's not working.
Either generic solution that will cover any rule, or none. Yes, it's bigger to implement, but it's not a hack.

**Comment by SpacePossum** [comment]
IMHO what would be nice on PSR-5 is a `@ignore [some namespace]\[some detail]`
Than all kinds of tooling good be notified that some violation should be ignore... than again the state of PSR-5 is maybe not in such a state to start introducing features like this...

**Comment by Wirone** [comment]
Reproduced on `3.16-DEV` (current master), so I'm going to keep it, even though I believe it's pretty hard to improve.

**Comment by Wirone** [comment]
I am going to close this one, because in terms of `StaticLambdaFixer` it was already pointed out that this particular riskiness mentioned by @Slamdunk is documented in the rule itself. When it comes to ignoring parts of code from being fixed, the #4512 fits better.
